### PR TITLE
KAFKA-9726 IdentityReplicationPolicy

### PR DESCRIPTION
**Pull request by ryannedolan**
Enable active/passive, one-way replication without renaming topics, similar to MM1. This implementation is described in KIP-382 (adopted), originally as "LegacyReplicationPolicy".

This enables operators to migrate from MM1 to MM2 without re-architecting their replication flows, and enables some additional use-cases for MM2. For example, operators may wish to "upgrade" their Kafka clusters by mirroring everything to a completely new cluster. Such a migration would have been difficult with either MM1 or MM2 previously.

When using IdentityReplicationPolicy, operators should be aware that MM2 will not be able to detect cycles among replicated topics. A misconfigured topology may result in replicating the same records back-and-forth or in an infinite loop. However, we don't prevent this behavior, as some use-cases involve filtering records (via SMTs) to prevent cycles.

This PR includes major contributions from @mdedetrich and @ivanyu, so please include them in the commit log!

### CODE REVIEW

File: connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java [hunk 1]
```diff
@@ -171,14 +177,22 @@ public void startClusters() throws Exception {
         waitForTopicCreated(primary, "mm2-status.backup.internal");
         waitForTopicCreated(primary, "mm2-offsets.backup.internal");
         waitForTopicCreated(primary, "mm2-configs.backup.internal");
-        
+ 
+        waitForTopicCreated(backup, "mm2-status.primary.internal");
+        waitForTopicCreated(backup, "mm2-offsets.primary.internal");
+        waitForTopicCreated(backup, "mm2-configs.primary.internal");
+       
         backup.start();
```
Comment (ryannedolan) [inline connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationBaseTest.java:188]: whoops, bug here!

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java [hunk 2]
```diff
@@ -60,7 +60,7 @@
     private ReplicationPolicy replicationPolicy;
     private Map<String, Object> consumerConfig;
 
-    public MirrorClient(Map<String, Object> props) {
+    public MirrorClient(Map<String, ?> props) {
```
Comment (ryannedolan) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java]: This is an incidental API change, but I don't _think_ it is a breaking change.

**Comment by mdedetrich** [comment]
@ryannedolan Since https://github.com/apache/kafka/pull/10762 was merged maybe it makes sense to rebase against the current `trunk`? Some of the tests in this PR have assert statements without the failure messages which have been just been fixed (I believe you can just copy some of those assert failure messages to make sure its consistent).

**Comment by mdedetrich** [comment]
@mimaison As my PR at https://github.com/apache/kafka/pull/10648 has been superseded by this one, would it be possible to do a quick review of it? Afaik its ready and @ryannedolan has done great work in avoiding the need for a KIP so there shouldn't be any blockers.

**Comment by ryannedolan** [comment]
None of the failing tests are related. Ready to merge.

**Comment by mimaison** [comment]
@ryannedolan @mdedetrich Thanks, I'll try to take a look this week

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java (hunk 2 above)
Comment (mimaison) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java]: Is this actually needed?

File: connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java [hunk 3]
```diff
@@ -159,4 +191,12 @@ public void remoteTopicsSeparatorTest() throws InterruptedException {
         assertTrue(remoteTopics.contains("source3__source4__source5__topic6"));
     }
 
+    public void testIdentityReplicationTopicSource() {
```
Comment (mimaison) [inline connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java:198]: Missing `@Test` annotation

File: connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java [hunk 4]
```diff
@@ -159,4 +191,12 @@ public void remoteTopicsSeparatorTest() throws InterruptedException {
...
+        MirrorClient client = new FakeMirrorClient(
+            new IdentityReplicationPolicy("primary"), Arrays.asList());
+        assertEquals("topic1", client.replicationPolicy()
+            .formatRemoteTopic("primary", "topic1"));
```
Comment (mimaison) [inline connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java:202]: Should we also try `formatRemoteTopic()` with a heartbeat topic?

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java [hunk 5]
```diff
@@ -492,7 +492,12 @@ boolean isCycle(String topic) {
         } else if (source.equals(sourceAndTarget.target())) {
             return true;
         } else {
-            return isCycle(replicationPolicy.upstreamTopic(topic));
+            String upstreamTopic = replicationPolicy.upstreamTopic(topic);
+            if (upstreamTopic.equals(topic)) {
```
Comment (mimaison) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:496]: Can we cover this new branch with a test in `MirrorSourceConnectorTest`?

**Review by mimaison — CHANGES_REQUESTED** [review]
Thanks @ryannedolan for the PR. It looks good overall, I left a few small comments

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java (hunk 2 above)
Comment (ryannedolan) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/MirrorClient.java]: i thought it was but i must have fixed something cuz it's compiling fine without it now :)

File: connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java (hunk 3 above)
Comment (ryannedolan) [reply connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java:198]: fixed thanks

File: connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java (hunk 4 above)
Comment (ryannedolan) [reply connect/mirror-client/src/test/java/org/apache/kafka/connect/mirror/MirrorClientTest.java:202]: good call, thanks! fixed.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java (hunk 5 above)
Comment (ryannedolan) [reply connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:496]: added testIdentityReplication(), thanks

**Comment by ryannedolan** [comment]
@mimaison thanks for the feedback and suggestions. I've updated accordingly.

File: connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/IdentityReplicationPolicy.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    public static final String SOURCE_CLUSTER_ALIAS_CONFIG = "source.cluster.alias";
+
+    private String sourceClusterAlias = null;
+
+    public IdentityReplicationPolicy() {
+        //nop
+    }
+
+    // Visible for testing
+    IdentityReplicationPolicy(String sourceClusterAlias) {
```
Comment (mimaison) [inline connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/IdentityReplicationPolicy.java]: Not a big deal but do we need this extra constructor? Can we instead use the public constructor and then `configure()`?

Comment (ryannedolan) [reply connect/mirror-client/src/main/java/org/apache/kafka/connect/mirror/IdentityReplicationPolicy.java]: fixed, thanks

**Comment by mdedetrich** [comment]
@ryannedolan I actually just realized, does it make sense to add the `IdentityReplicationPolicy` to the release notes so that people are aware of it (talking about `docs/streams/upgrade-guide.html`)

**Review by mimaison — APPROVED** [review]

**Comment by mimaison** [comment]
Yes, we can add a line in https://github.com/apache/kafka/blob/trunk/docs/upgrade.html#L85 to introduce this new ReplicationPolicy

**Comment by mdedetrich** [comment]
Is there anything left on this PR to be merged (apart from the changelog which is a nice to have)?

**Comment by ryannedolan** [comment]
I think changelog and documentation updates can come after this is merged.

**Comment by mimaison** [comment]
Let's put a line in the changelog now so we're sure it's included in the release notes. I'm happy to merge once this is done.

I agree the documentation can come later.

File: docs/upgrade.html [hunk 7]
```diff
@@ -80,7 +80,13 @@ <h5><a id="upgrade_300_notable" href="#upgrade_300_notable">Notable changes in 3
         understood by brokers or version 2.5 or higher, so you must upgrade your kafka cluster to get the stronger semantics. Otherwise, you can just pass
         in <code>new ConsumerGroupMetadata(consumerGroupId)</code> to work with older brokers. See <a href="https://cwiki.apache.org/confluence/x/zJONCg">KIP-732</a> for more details.
     </li>
-
+    <li> The Connect-based MirrorMaker (MM2) includes changes to support <code>IdentityReplicationPolicy</code>, enabling replication without renaming topics.
```
Comment (ryannedolan) [inline docs/upgrade.html:86]: @mimaison @mdedetrich wdyt?

Comment (mdedetrich) [reply docs/upgrade.html:86]: I would make a quick amendment stating that it works like the original MM1 to make it more clear.

**Comment by mimaison** [comment]
@ryannedolan I was about to merge but I noticed `IdentityReplicationIntegrationTest.testOffsetSyncsTopicsOnTarget()` is failing. It probably needs some adjustements to work with this policy as the computed offset-syncs topic name looks odd:
```
org.opentest4j.AssertionFailedError: Condition not met within timeout 30000. Unable to find checkpoints for primarytest-topic-1 ==> expected: <true> but was: <false>
```

File: docs/upgrade.html (hunk 7 above)
Comment (mimaison) [reply docs/upgrade.html:86]: Yes I think it's enough, thanks

**Comment by ryannedolan** [comment]
> testOffsetSyncsTopicsOnTarget

Yep good catch. Fixed!

**Comment by mimaison** [comment]
Thanks @ryannedolan, @mdedetrich and @ivanyu

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java [hunk 8]
```diff
@@ -175,6 +175,7 @@ public MirrorClientConfig clientConfig(String cluster) {
         props.putAll(stringsWithPrefix("header.converter"));
         props.putAll(stringsWithPrefix("task"));
         props.putAll(stringsWithPrefix("worker"));
+        props.putAll(stringsWithPrefix("replication.policy"));
```
Comment (Justinwins) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorMakerConfig.java:178]: In MirrorClientConfig ,it seems that it's not necessary to add replication.policy  into the props.
As i see , org.apache.kafka.connect.mirror.MirrorClientConfig#replicationPolicy  initialise the ReplicationPolicy  instance,
and  in  MirrorClient this instance is used ,where ReplicationPolicy   takes effect for real.

File: docs/upgrade.html [hunk 9]
```diff
@@ -83,7 +83,13 @@ <h5><a id="upgrade_300_notable" href="#upgrade_300_notable">Notable changes in 3
...
+        The existing <code>DefaultReplicationPolicy</code> is still used by default, but identity replication can be enabled via the
+        <code>replication.policy</code> configuration property. This is especially useful for users migrating from the older MirrorMaker (MM1), or for
```
Comment (Justinwins) [inline docs/upgrade.html:88]: i think it's more clear to  say  "replication.policy.class "  here ,you know ,  which means it's configured in that form in mm2.properties  file.

Friendly to beginners .

**Pull request merged**